crowbar_framework/app/models/service_object.rb

```diff
@@ -1826,4 +1821,71 @@ def skip_unready_nodes(bc, inst, new_elements, old_elements)
     logger.debug("skip_unready_nodes: exit for #{bc}:#{inst}")
     [cleaned_elements, pre_cached_nodes]
   end
+
+  # try to know if we can skip a node from running chef-client
+  def skip_unchanged_node?(node_name, old_role, new_role)
+    # get the list of barclamps where this method will run
+    skip_unchanged_nodes_barclamps = Rails.application.config.experimental.fetch(
+      "skip_unchanged_nodes", {}
+    ).fetch("barclamps", false)
+
+    # Check if barclamp is included in the list, if not skip
+    return false unless skip_unchanged_nodes_barclamps.include?(@bc_name)
+
+    @logger.debug("#{@bc_name} skip_batch_for_node? entering: #{node_name}")
+
+    # if old_role is nil, then we are applying the barclamp for the first time
+    return false if old_role.nil?
+
+    # if attributes have changed, we need to run
+    return false if node_changed_attributes?(node_name, old_role, new_role)
+
+    # if the node changed roles, then we need to apply
+    if node_changed_roles?(node_name, old_role, new_role, role_constraints.keys)
+      @logger.debug("#{@bc_name} skip_batch_for_node?: not skipping #{node_name} (changed roles)")
+      return false
+    end
+
+    # by this point its safe to assume that we can skip the node as nothing has changed on it
+    # same attributes, same roles so skip it
+    @logger.debug("#{@bc_name} skip_batch_for_node? skipping: #{node_name}")
+    @logger.debug("#{@bc_name} skip_batch_for_node? exiting: #{node_name}")
+    true
+  end
+
+  # return true if the new attributes are different from the old ones
+  def node_changed_attributes?(node, old_role, new_role)
+    if old_role.default_attributes[@bc_name] != new_role.default_attributes[@bc_name]
+      logger.debug("#{@bc_name} skip_batch_for_node?: not skipping #{node} (attribute change)")
+      return true
+    end
+
+    false
+  end
+
+  # return true if the node has changed roles
+  def node_changed_roles?(node, old_role, new_role, role_list)
+    # Check if some role of this barclamp has changed in the node
+    role_list.each do |role_name|
```

Comment:
You don't need to pass a `role_list`, you have this information in `{old,new}_role.override_attributes[@bc_name]["elements"]` already (the roles are the keys of these two hashes)